File: go/deck.go
package main

import (

	"fmt"
	"math/rand"
	"sort"
	"strings"
	"time"

)

////////////// CARD CLASS

// card represents a single card
type Card struct {
	Face string
	Suit string
}

// string returns the string rep of a card
func (c Card) String() string {
	return c.Face + c.Suit
}

// deck reps a deck of cards and game state
type Deck struct {

	cards		[][]Card
	dealtHands	[]string
	results		[]*Player
	isTest		bool

}

type HandType int

// global consts
const (

	HighCard HandType = iota + 1
	Pair
	TwoPair
	ThreeOfAKind
	Straight
	Flush
	FullHouse
	FourOfAKind
	StraightFlush
	RoyalStraightFlush

)

func (h HandType) String() string {

	switch h {

	case HighCard:
		return "High Card"
	case Pair:
		return "Pair"
	case TwoPair:
		return "Two Pair"
	case ThreeOfAKind:
		return "Three of a Kind"
	case Straight:
		return "Straight"
	case Flush:
		return "Flush"
	case FullHouse:
		return "Full House"
	case FourOfAKind:
		return "Four of a Kind"
	case StraightFlush:
		return "Straight Flush"
	case RoyalStraightFlush:
		return "Royal Straight Flush"
	default:
		return "Unknown"

	}

}

/////////////////// PLAYER CLASS

// player class that reps a player's hand and its eval
type Player struct {

	hand		string
	wordList	[]string
	order		int
	handType	HandType
	suitCount	map[string]int
	rankCount	map[int]int

}


// new player instance
func NewPlayer(hand string) *Player {

	p := &Player{

		hand:		hand,
		suitCount:	make(map[string]int),
		rankCount:	make(map[int]int),

	}
	p.convertHand()
	return p

}

// convert hand to card counts
func (p *Player) convertHand() {

	p.wordList = strings.Fields(p.hand)

	for _, word := range p.wordList {

		faceStr := word[:len(word)-1]
		suitChar := word[len(word)-1:]

		p.suitCount[suitChar]++
		faceVal := faceToValue(faceStr)
		p.rankCount[faceVal]++

	}

	p.evaluateHand()

}

// evaluate which hand
func (p *Player) evaluateHand() {

	if p.isFlush() && p.isStraight() && p.isRoyal() {

		p.handType = RoyalStraightFlush
		p.order = 1

	}else if p.isFlush() && p.isStraight() {

		p.handType = StraightFlush
		p.order = 2

	}else if p.isFourOfAKind(){

		p.handType = FourOfAKind
		p.order = 3

	}else if p.isFullHouse() {

		p.handType = FullHouse
		p.order = 4

	}else if p.isFlush() {

		p.handType = Flush
		p.order = 5

	}else if p.isStraight() {

		p.handType = Straight
		p.order = 6

	}else if p.isThreeOfAKind() {

		p.handType = ThreeOfAKind
		p.order = 7

	}else if p.isTwoPair() {

		p.handType = TwoPair
		p.order = 8

	}else if p.isPair() {

		p.handType = Pair
		p.order = 9

	}else {

		p.handType = HighCard
		p.order = 10

	}

}

/////////// HELPER FUNCS
func (p *Player) isFlush() bool {

	for _, count := range p.suitCount {

		if count == 5 {

			return true

		}

	}

	return false

}

func (p *Player) isStraight() bool {

	var ranks []int
	for rank := range p.rankCount {

		ranks = append(ranks, rank)

	}
	sort.Ints(ranks)

	//check for low ace straight
	if len(ranks) == 5 && ranks[4] == 14 && ranks[0] == 2 && ranks[1] == 3 && ranks[2] == 4 && ranks[3] == 5 {

		return true

	}

	// check for reg straight
	if len(ranks) != 5 {

		return false

	}
	for i := 1; i < len(ranks); i++ {

		if ranks[i] != ranks[i-1]+1 {

			return false

		}

	}

	return true

}

func (p *Player) isRoyal() bool {

	required := []int{10, 11, 12, 13, 14}
	for _, rank := range required {

		if p.rankCount[rank] == 0 {

			return false

		}

	}

	return true

}

func (p *Player) isFourOfAKind() bool {

	for _, count := range p.rankCount {

		if count == 4 {

			return true

		}

	}

	return false

}


func (p *Player) isFullHouse() bool {

	hasThree := false
	hasPair := false
	for _, count := range p.rankCount {

		if count == 3 {

			hasThree = true

		} else if count == 2 {

			hasPair = true

		}

	}

	return hasThree && hasPair

}

func (p *Player) isThreeOfAKind() bool {

	for _, count := range p.rankCount {

		if count == 3 {

			return true

		}

	}


	return false
}

func (p *Player) isTwoPair() bool {

	pairCount := 0
	for _, count := range p.rankCount {

		if count == 2 {

			pairCount++

		}

	}

	return pairCount == 2

}

func (p *Player) isPair() bool {

	for _, count := range p.rankCount {

		if count == 2 {

			return true

		}

	}

	return false

}


//////////////////// DECK CLASS

// returns highest value of card not in a pair
func (p *Player) GetHighCard() int {

	highCard := -1
	for rank, count := range p.rankCount {

		if count == 1 && rank > highCard {

			highCard = rank

		}

	}

	return highCard

}

// returns highest card value 
func (p *Player) GetKicker() int {

	kicker := -1
	for rank, count := range p.rankCount {

		if count == 1 && rank > kicker {

			kicker = rank

		}

	}

	return kicker

}

// returns ranks of the two pairs
func (p *Player) GetTwoPairRanks() []int {

	var pairs []int
	for rank, count := range p.rankCount {

		if count == 2 {

			pairs = append(pairs, rank)

		}

	}

	sort.Sort(sort.Reverse(sort.IntSlice(pairs)))
	return pairs

}

// returns rank of 3 or 4 of a kind
func (p *Player) GetSetRank() int {

	for rank, count := range p.rankCount {

		if count >= 3 {

			return rank

		}

	}

	return -1

}

// create new deck
func NewDeck(isTest bool) *Deck {

	d := &Deck{

		cards:		make([][]Card, 4),
		dealtHands:	make([]string, 6),
		results:	make([]*Player, 6),
		isTest:		isTest,

	}

	if !isTest {

		suits := []string{"D", "C", "H", "S"}
		faces := []string{"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"}

		for i := range d.cards {

			d.cards[i] = make([]Card, 13)
			for j := range d.cards[i] {

				d.cards[i][j] = Card{Face: faces[j], Suit: suits[i]}

			}

		}

	}

	return d

}

// shuffle deck!
func (d *Deck) Shuffle() {

	rand.Seed(time.Now().UnixNano())

	// flatten the deck b4 shuffling
	flat := make([]Card, 52)
	idx := 0
	for i := range d.cards{

		for j := range d.cards[i] {

			flat[idx] = d.cards[i][j]
			idx++

		}

	}

	// rand
	for i := len(flat) - 1; i > 0; i-- {

		j := rand.Intn(i + 1)
		flat[i], flat[j] = flat[j], flat[i]

	}

	// restore
	idx = 0
	for i := range d.cards {

		for j := range d.cards[i] {

			d.cards[i][j] = flat[idx]
			idx++

		}

	}


}

// prints the shuffled deck
func (d *Deck) PrintDeck() {

	fmt.Println("*** Shuffled 52 card deck: ")
	for i := range d.cards {

		for j := range d.cards[i] {

			fmt.Printf("%s ", d.cards[i][j])

		}

		fmt.Println()

	}

}

// deal hands
func (d *Deck) DealHands() {

	hands := make([]string, 6)
	count := 0

	for i := range d.cards {

		for j := range d.cards[i] {

			if count < 30 {

				handIndex := count % 6
				hands[handIndex] += fmt.Sprintf("%s ", d.cards[i][j])
				count++

			}

		}

	}

	d.dealtHands = hands
}

// rank hands
func (d *Deck) RankHands(hands []string) {

	for i, hand := range hands {

		d.results[i] = NewPlayer(hand)

	}

	d.tieBreakers()

}

// bleh tiebreakers function
func (d *Deck) tieBreakers() {

	sort.Slice(d.results, func(i, j int) bool {

		if d.results[i].order != d.results[j].order {

			return d.results[i].order < d.results[j].order

		}

		// resoolve ties based on hand type
		switch d.results[i].handType {

		case RoyalStraightFlush:
			return false
		case StraightFlush, Straight, Flush:
			return d.results[i].GetHighCard() > d.results[j].GetHighCard()
		case FourOfAKind:
			if rankI, rankJ := d.results[i].GetSetRank(), d.results[j].GetSetRank(); rankI != rankJ {

				return rankI > rankJ

			}
			return d.results[i].GetKicker() > d.results[j].GetKicker()
		case FullHouse, ThreeOfAKind:
			return d.results[i].GetSetRank() > d.results[j].GetSetRank()
		case TwoPair:
			pairsI := d.results[i].GetTwoPairRanks()
			pairsJ := d.results[j].GetTwoPairRanks()
			if pairsI[0] != pairsJ[0] {

				return pairsI[0] > pairsJ[0]

			}
			if pairsI[1] != pairsJ[1] {

				return pairsI[1] > pairsJ[1]

			}
			return d.results[i].GetKicker() > d.results[j].GetKicker()
		case Pair:
			if rankI, rankJ := d.results[i].GetSetRank(), d.results[j].GetSetRank(); rankI != rankJ {

				return rankI > rankJ

			}
			return d.results[i].GetKicker() > d.results[j].GetKicker()
		case HighCard:
			return d.results[i].GetHighCard() > d.results[j].GetHighCard()
		}

		return false
	})

}

// prints 6 habds
func (d *Deck) PrintHands() {

	fmt.Println("*** Here are the six hands...")
	for _, hand := range d.dealtHands {

		fmt.Println(hand)

	}

}

// prints remaining cards
func (d *Deck) PrintRemainingCards() {

	fmt.Println("*** Here is what remains in the deck...")
	count := 0
	for i := range d.cards {

		for j := range d.cards[i] {

			if count >= 30 {

				fmt.Printf("%s ", d.cards[i][j])

			}
			count++

		}

	}

	fmt.Println()
}

// print final results
func (d *Deck) PrintResults() {

	fmt.Println("\n--- WINNING HAND ORDER ---")
	for _, player := range d.results {

		fmt.Printf("%s - %s\n", strings.Join(player.wordList, " "), player.handType)

	}

	fmt.Println()
}

// convert face to value
func faceToValue(face string) int {

	switch face {

	case "J":
		return 11
	case "Q":
		return 12
	case "K":
		return 13
	case "A":
		return 14
	default:
		val := 0
		fmt.Sscanf(face, "%d", &val)
		return val

	}

}

// retrieve deck from gamemode
func (d *Deck) RetrieveDeck(mode interface{ GetTestArray() []string })  {

	d.dealtHands = mode.GetTestArray()

}

File: rust/src/deck.rs

use rand::seq::SliceRandom;  // For shuffling functionality
use rand::thread_rng;        // For random number generation
use crate::card::Card;       // Card struct import
use crate::player::Player;   // Player struct import
use crate::game::GameMode;   // GameMode struct import

// Represents a deck of 52 playing cards arranged in a 2D array (4x13)
pub struct Deck {
    deck_array: Vec<Vec<Card>>,   // 2D vector storing cards (4 suits x 13 faces)
    dealt_cards: Vec<String>,     // Stores the six hands that are dealt
    result: Vec<Player>,          // Stores player results for ranking
    rows: usize,                  // Number of rows (4 for suits)
    cols: usize,                  // Number of columns (13 for faces)
}

impl Deck {
    // Creates a new deck instance. If test=true, creates empty deck for testing
    pub fn new(test: bool) -> Self {
        let rows = 4;
        let cols = 13;
        let suits = vec!['D', 'C', 'H', 'S'];
        let faces = vec!["A", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"];

        // Initialize empty deck array
        let mut deck_array = vec![vec![Card::new(' ', String::new()); cols]; rows];

        // If not in test mode, populate deck with cards
        if !test {
            for i in 0..rows {
                for j in 0..cols {
                    deck_array[i][j] = Card::new(
                        suits[i],
                        faces[j].to_string()
                    );
                }
            }
        }

        Deck {
            deck_array,
            dealt_cards: vec![String::new(); 6],  // Space for 6 hands
            result: Vec::new(),
            rows,
            cols,
        }
    }

    // Creates a new empty test deck
    pub fn new_test() -> Self {
        Self::new(true)
    }

    // Shuffles the deck using random number generation
    pub fn shuffle(&mut self) {
        let mut rng = thread_rng();
        // Convert 2D array to 1D for shuffling
        let mut flat_deck: Vec<Card> = self.deck_array
            .iter()
            .flat_map(|row| row.iter().cloned())
            .collect();

        flat_deck.shuffle(&mut rng);

        // Convert back to 2D array
        let mut loc = 0;
        for i in 0..self.rows {
            for j in 0..self.cols {
                self.deck_array[i][j] = flat_deck[loc].clone();
                loc += 1;
            }
        }
    }

    // Prints the entire shuffled deck
    pub fn print_shuffled(&self) {
        println!("*** Shuffled 52 card deck: ");
        for row in &self.deck_array {
            for card in row {
                print!("{} ", card.to_string());
            }
            println!();
        }
    }

    // Deals six hands of 5 cards each from the deck
    pub fn deal_hands(&mut self) {
        let mut hands = vec![String::new(); 6];
        let mut count = 0;

        'outer: for i in 0..self.rows {
            for j in 0..self.cols {
                if count < 30 {  // Deal 30 cards total (6 hands * 5 cards)
                    let hand_index = count % 6;
                    hands[hand_index].push_str(&self.deck_array[i][j].to_string());
                    hands[hand_index].push(' ');
                    count += 1;
                }else {
                    break 'outer;
                }
            }
        }

        self.dealt_cards = hands;
    }

    // Prints all six dealt hands
    pub fn print_hands(&self) {
        println!("*** Here are the six hands...");
        for hand in &self.dealt_cards {
            println!("{}", hand);
        }
    }

    // Prints the remaining cards in the deck after dealing
    pub fn print_remaining(&mut self) {
        println!("*** Here is what remains in the deck... ");
        let mut count = 0;

        for row in &self.deck_array {
            for card in row {
                if count >= 30 {  // Print only cards after the first 30 dealt
                    print!("{} ", card.to_string());
                }
                count += 1;
            }
        }

        println!();

        // Rank the hands after showing remaining cards
        let dealt_copy: Vec<String> = self.dealt_cards.iter().cloned().collect();
        self.rank_hands(&dealt_copy);
    }

    // Ranks all hands by poker hand value
    pub fn rank_hands(&mut self, hands: &[String]) {
        self.result.clear();
        for hand in hands {
            if !hand.is_empty() {
                let mut player = Player::new(hand.clone());
                player.set_final_scores();
                self.result.push(player);
            }
        }
    }

    // For test mode: retrieves test deck from GameMode
    pub fn retrieve_deck(&mut self, mode: &GameMode) {
        self.dealt_cards = mode.test_array.clone();
    }

    // Prints final results after ranking all hands
    pub fn print_results(&mut self) {
        self.sort_order();
        self.tie_breakers();
        println!();
        println!("--- WINNING HAND ORDER ---");
        for player in &self.result {
            println!("{} - {}", player.word_list.join(" "), player.hand);
        }
        println!();
    }

    // Sorts hands by their poker hand ranking
    fn sort_order(&mut self) {
        self.result.sort_by_key(|player| player.order);
    }

    // Applies tie-breaker rules based on hand type
    fn tie_breakers(&mut self) {
        for i in 0..self.result.len() {
            for j in (i + 1)..self.result.len() {
                if self.result[i].order == self.result[j].order {
                    match self.result[i].hand.as_str() {
                        "Straight Flush" | "Flush" | "Royal Straight Flush" => {
                            self.compare_high_cards(i, j);
                        } "Straight" => {
                            self.compare_high_cards(i, j);
                        } "Two Pair" => {
                            self.compare_two_pairs(i, j);
                        } "Pair" => {
                            self.compare_pairs(i, j);
                        } "High Card" => {
                            self.compare_high_cards_and_kicker(i, j);
                        } "Three of a Kind" => {
                            self.compare_three_of_kind(i, j);
                        } "Four of a Kind" => {
                            self.compare_four_of_kind(i, j);
                        } "Full House" => {
                            self.compare_full_house(i, j);
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    //////////// Tie-breaker helper functions

    // Compares hands based on highest card value
    fn compare_high_cards(&mut self, i: usize, j: usize) {
        let high_i = self.result[i].get_high_card();
        let high_j = self.result[j].get_high_card();

        if high_i < high_j {
            self.result.swap(i, j);
        }
    }
    
    // Compares hands based on high card and kicker
    fn compare_high_cards_and_kicker(&mut self, i: usize, j: usize) {
        let high_i = self.result[i].get_high_card();
        let high_j = self.result[j].get_high_card();

        if high_i < high_j {
            self.result.swap(i, j);
        }else if high_i == high_j {
            let kicker_i = self.result[i].get_kicker();
            let kicker_j = self.result[j].get_kicker();
            if kicker_i < kicker_j {
                self.result.swap(i, j);
            }
        }
    }

    // Compares pairs based on suit ranking
    fn compare_pairs(&mut self, i: usize, j: usize) {
        let rank_i = Card::rank_suits(self.result[i].get_high_card_suit());
        let rank_j = Card::rank_suits(self.result[j].get_high_card_suit());
        let suit_i = self.result[i].get_high_card_suit();
        let suit_j = self.result[j].get_high_card_suit();

        if rank_i < rank_j {
            self.result.swap(i, j);
        } else if rank_i == rank_j && suit_i < suit_j {
            self.result.swap(i, j);
        }
    }

    // Compares two pair hands
    fn compare_two_pairs(&mut self, i: usize, j: usize) {
        let ranks_i = self.result[i].get_to_pair_ranks();
        let ranks_j = self.result[j].get_to_pair_ranks();
        
        if ranks_i < ranks_j {
            self.result.swap(i, j);
        } else if ranks_i == ranks_j {
            let suit_i = self.result[i].get_kicker_suit();
            let suit_j = self.result[j].get_kicker_suit();
            if Card::face_to_value(&suit_i.to_string()) < Card::face_to_value(&suit_j.to_string()) {
                self.result.swap(i, j);
            }
        }
    }

    // Compares three of a kind hands
    fn compare_three_of_kind(&mut self, i: usize, j: usize) {
        let rank_i = self.result[i].get_toak_rank();
        let rank_j = self.result[j].get_toak_rank();
        
        if rank_i < rank_j {
            self.result.swap(i, j);
        } else if rank_i == rank_j {
            let kicker_i = self.result[i].get_kicker();
            let kicker_j = self.result[j].get_kicker();
            if kicker_i < kicker_j {
                self.result.swap(i, j);
            }
        }
    }

    // Compares four of a kind hands
    fn compare_four_of_kind(&mut self, i: usize, j: usize) {
        let rank_i = self.result[i].get_foak_rank();
        let rank_j = self.result[j].get_foak_rank();
        
        if rank_i < rank_j {
            self.result.swap(i, j);
        } else if rank_i == rank_j {
            let kicker_i = self.result[i].get_kicker();
            let kicker_j = self.result[j].get_kicker();
            if kicker_i < kicker_j {
                self.result.swap(i, j);
            }
        }
    }

    // Compares full house hands based on three of a kind value
    fn compare_full_house(&mut self, i: usize, j: usize) {
        let rank_i = self.result[i].get_toak_rank();
        let rank_j = self.result[j].get_toak_rank();
        
        if rank_i < rank_j {
            self.result.swap(i, j);
        }
    }
}


